packages/synapse-react-client/src/synapse-client/SynapseClient.ts

```diff
@@ -2099,54 +2105,57 @@ export const uploadFile = (
           fileUploadReject,
           progressCallback,
           getIsCancelled,
+          abortController,
         )
       })
     },
   )
 }
 
-export const calculateMd5 = (fileBlob: File | Blob): Promise<string> => {
+/**
+ * Calculate the MD5 of the data in a Blob. This function is memoized, so if the same {@link Blob} object is
+ * passed after the MD5 is computed, no computation will occur.
+ */
+export const calculateMd5: (blob: Blob) => Promise<string> = memoize(blob => {
   // code taken from md5 example from library
   return new Promise((resolve, reject) => {
-    const blobSlice = File.prototype.slice,
-      file = fileBlob,
-      chunkSize = 2097152, // Read in chunks of 2MB
-      chunks = Math.ceil(file.size / chunkSize),
-      spark = new SparkMD5.ArrayBuffer(),
-      fileReader = new FileReader()
+    const chunkSize = 2097152 // Read in chunks of 2M
+    const chunks = Math.ceil(blob.size / chunkSize)
+    const spark = new SparkMD5.ArrayBuffer()
+    const fileReader = new FileReader()
     let currentChunk = 0
 
     fileReader.onload = function (e) {
-      console.log('read chunk nr', currentChunk + 1, 'of', chunks)
+      console.debug('read chunk nr', currentChunk + 1, 'of', chunks)
       spark.append(fileReader.result as ArrayBuffer) // Append array buffer
       currentChunk++
 
       if (currentChunk < chunks) {
         loadNext()
       } else {
-        console.log('finished loading')
+        console.debug('finished loading')
         const md5: string = spark.end()
-        console.info('computed hash', md5) // Compute hash
+        console.debug('computed hash', md5) // Compute hash
         resolve(md5)
       }
     }
 
     fileReader.onerror = function () {
-      console.warn('oops, something went wrong.')
+      console.warn('oops, something went wrong.', fileReader.error)
       reject(fileReader.error)
     }
 
     const loadNext = () => {
       const start = currentChunk * chunkSize,
-        end = start + chunkSize >= file.size ? file.size : start + chunkSize
+        end = start + chunkSize >= blob.size ? blob.size : start + chunkSize
 
-      fileReader.readAsArrayBuffer(blobSlice.call(file, start, end))
+      fileReader.readAsArrayBuffer(blob.slice(start, end))
     }
     loadNext()
   })
-}
+})
 
-const processFilePart = (
+const processFilePart = async (
```

Comment:
Rewritten to use async/await instead of Promise.then. No logical changes other than passing `getIsCancelled` (missing) and `AbortController` (new) through to the recursive call on retry.